test/e2e/framework/service/jig.go

```diff
@@ -260,21 +261,41 @@ func (j *TestJig) CreateLoadBalancerService(timeout time.Duration, tweak func(sv
 // GetEndpointNodes returns a map of nodenames:external-ip on which the
 // endpoints of the Service are running.
 func (j *TestJig) GetEndpointNodes() (map[string][]string, error) {
-	nodes, err := e2enode.GetBoundedReadySchedulableNodes(j.Client, MaxNodesForEndpointsTests)
+	nodes, err := j.ListNodesWithEndpoint()
 	if err != nil {
 		return nil, err
 	}
-	epNodes, err := j.GetEndpointNodeNames()
+	nodeMap := map[string][]string{}
+	for _, node := range nodes {
+		nodeMap[node.Name] = e2enode.GetAddresses(&node, v1.NodeExternalIP)
+	}
+	return nodeMap, nil
+}
+
+// ListNodesWithEndpoint returns a list of nodes on which the
+// endpoints of the given Service are running.
+func (j *TestJig) ListNodesWithEndpoint() ([]v1.Node, error) {
+	nodeNames, err := j.GetEndpointNodeNames()
 	if err != nil {
 		return nil, err
 	}
-	nodeMap := map[string][]string{}
-	for _, n := range nodes.Items {
-		if epNodes.Has(n.Name) {
-			nodeMap[n.Name] = e2enode.GetAddresses(&n, v1.NodeExternalIP)
-		}
+	ctx := context.TODO()
+	namesList := make([]string, 0, nodeNames.Len())
+	for name := range nodeNames {
+		namesList = append(namesList, name)
 	}
-	return nodeMap, nil
+	req, err := labels.NewRequirement(v1.LabelHostname, selection.In, namesList)
```

Comment:
FTR - that will usually work, unless you exceed the allowed length for url params (which isn't that long IIRC).

I guess I wasn't clear, but what I meant was:
- list *all* nodes in the cluster
- do client-side filtering for nodeName you want

so sth like
```
allNodes, err := j.Client.CoreV1().Nodes().List(ctx, metav1.ListOptions{}))
if err != nil {
  return nil, err
}

nodes := make([]v1.Node, 0, nodeName.Len())
for i := range allNodes {
  node := &allNodes[i]
  if !nodeNames.Has(node.Name) {
    continue
  }
  node = append(nodes, *node)
}
```

Reply:
yeah I understand what you mean, but if we do client-side filtering we may have ~5000 iterations in the worst case. Isn't it something we try to avoid?

Reply:
The same would happen server-side if you do it using the selector.
Iterating is cheap. The only potential concern is listing and sending all those data, but we do that in many other tests anyway.

So no - it's fine to do that.

Reply:
updated. please check again.